kvbc/src/v4blockchain/detail/keys_history.cpp

```diff
@@ -0,0 +1,226 @@
+                                        const BlockId block_id,
+                                        concord::storage::rocksdb::NativeWriteBatch& write_batch) {
+  const auto& prefix = category_mapping_.categoryPrefix(category_id);
+  std::string key_prefix;
+  for (const auto& k : updates) {
+    auto itr = native_client_->getIterator(v4blockchain::detail::KEYS_HISTORY_CF);
+    key_prefix = prefix;
+    key_prefix.append(k);
+    BlockId oldest_block_to_del = block_id;
+    std::string oldest_version_to_delete_key;
+    do {  // TODO (efrat): needed? is it only 1 block revert or possibly more at one call?
+      std::string block_key = v4blockchain::detail::Blockchain::generateKey(oldest_block_to_del);
+      oldest_version_to_delete_key = key_prefix;
+      oldest_version_to_delete_key.append(block_key);
+      itr.seekAtLeast(oldest_version_to_delete_key);
+
+      // TODO (efrat): if we revert only 1 last block we can check key including version. o/w, check prefix+k (?)
+      //      if (itr && itr.key().substr(0, itr.key().size() - sizeof(BlockId)) == key_prefix) {
+      if (itr && itr.key() == oldest_version_to_delete_key) {
+        write_batch.del(v4blockchain::detail::KEYS_HISTORY_CF, itr.key());
+        LOG_DEBUG(V4_BLOCK_LOG,
+                  "Revert found key " << k << " in version " << oldest_block_to_del << " category " << category_id
+                                      << " prefix " << prefix);
+        oldest_block_to_del++;
+      } else {
+        LOG_DEBUG(V4_BLOCK_LOG,
+                  "Couldn't find key " << k << " in version " << oldest_block_to_del << " category " << category_id
+                                       << " prefix " << prefix);
+        break;
+      }
+    } while (true);
+  }
+}
+
+std::optional<BlockId> KeysHistory::getSnapshotValue(const std::string& category_id,
+                                                     const std::string& key,
+                                                     const BlockId version) const {
+  const auto& prefix = category_mapping_.categoryPrefix(category_id);
+  auto version_key = v4blockchain::detail::Blockchain::generateKey(version);
+  std::string get_key = prefix;
+  get_key.append(key);
+  get_key.append(version_key);
+
+  auto itr = native_client_->getIterator(v4blockchain::detail::KEYS_HISTORY_CF);
+  itr.seekAtMost(get_key);
+  if (!itr) {
+    LOG_DEBUG(V4_BLOCK_LOG,
+              "Reading key " << std::hash<std::string>{}(key) << " not found at version " << version << " category_id "
+                             << category_id << " prefix " << prefix << " key is hex "
+                             << concordUtils::bufferToHex(key.data(), key.size()) << " raw key " << key);
+    return std::nullopt;
+  }
+  auto found_key = itr.key();
+  const size_t total_key_size = found_key.size();
+  auto user_key_with_prefix = found_key.substr(0, total_key_size - sizeof(BlockId));
+
+  if (user_key_with_prefix != prefix + key) {
```

Comment:
can it happen? should it be an assert?

Reply:
In case where the asked key has not defined yet at the given version we will get the closest key (prev key in lexical order). I guess it's a valid scenario so I didn't assert

Reply:
oh, got it, cool